Finished First Iteration of Subsystem Conditions

nrgill28 opened the pull request:
So this is an interesting one. I moved the conditions part of the IntakeSubsystem into it's own set of classes so that all subsystems can benefit from them without having to re-write that code.

To summarize the new system: 
- A subsystem creates it's own instance of SubsystemConditionManager
- Commands and other subsystems can create conditions for that manager
- The conditions are created through `SubsystemConditionManager.createCondition` and are automatically registered so no extra code is required.
- The command or other subsystem can now enable and disable it's condition like normal

For an example of how these work, I've modified the intake subsystem and the operator intake command to use this system.

This is very much a first iteration of the code, it's not perfect but I'd like to put it out there and see what you guys think before I go any further. Aside from the regular code review, I'm also looking for feedback on a couple extra things:  
- The naming of basically everything. I'm bad at names, won't pretend I'm not. If you have a better name for a variable or class LMK.
- The placement of the two files. I put the SubsystemConditionManager and SubsystemCondition classes in the subsystems folder mainly because there wasn't already a better place to put it; but since they're not subsystems themselves I think it would be better to move them somewhere else.

nrgill28 committed:
Finished First Iteration of Subsystem Conditions

georgetzavelas commented on src/main/java/frc/robot/commands/OperatorIntakeCommand.java:
```diff
@@ -2,25 +2,27 @@
-        intakeSubsystem.setConditionActive("buttonPressed", IntakeSubsystem.Condition.ONLY_TELEOP);
+        // Initialize the condition
+        condition = intakeSubsystem.getConditionManager()
+                .createCondition("buttonPressed", SubsystemCondition.ConditionOptions.TELEOP, false);
```
Can we make "buttonPressed" into a constant?

georgetzavelas commented on src/main/java/frc/robot/subsystems/SubsystemCondition.java:
```diff
@@ -0,0 +1,79 @@
+
+import java.util.EnumSet;
+
+public class SubsystemCondition {
```
How about `SubsystemOperation`?

georgetzavelas commented on src/main/java/frc/robot/subsystems/SubsystemCondition.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * A list of the possible options conditions can have
+     */
+    public enum ConditionOptions {
```
How about `State`?
Then when invoking this elsewhere it will look like `SubsystemOperation.State.TELEOP`

georgetzavelas commented on src/main/java/frc/robot/subsystems/SubsystemCondition.java:
```diff
@@ -0,0 +1,79 @@
+        // Enum sets of the combinations for the options
+        public static EnumSet<ConditionOptions> TELEOP = EnumSet.of(ENABLED_FOR_TELEOP);
+        public static EnumSet<ConditionOptions> AUTO = EnumSet.of(ENABLED_FOR_AUTO);
+        public static EnumSet<ConditionOptions> BOTH = EnumSet.of(ENABLED_FOR_TELEOP, ENABLED_FOR_AUTO);
```
BOTH is requires me to have more context into what this means. How about something more explicit like `TELEP_AND_AUTO`?

georgetzavelas commented on src/main/java/frc/robot/subsystems/SubsystemConditionManager.java:
```diff
@@ -0,0 +1,70 @@
+ */
+public class SubsystemConditionManager {
+    // A list of all the conditions for this condition manager
+    private ArrayList<SubsystemCondition> conditions = new ArrayList<>();
```
Thoughts on using a HashSet here so you avoid duplicates?

nrgill28 committed:
Renamed SubsystemCondition* -> SubsystemOperation*

And various other improvements:
SubsystemConditionManager's internal list is now a HashSet to avoid duplicates
Instead of each operation using a pre-defined set of options, they can now define when they are applicable themselves (Some default ones included)

nrgill28 committed:
Constants should be final

georgetzavelas reviewed (approved):
![youre-awesome](https://user-images.githubusercontent.com/3611823/73596775-6d512e80-44f3-11ea-8ab5-e1188c4de313.gif)

Zinka010 commented:
![thumbs-up](https://media.giphy.com/media/3o7abKhOpu0NwenH3O/giphy.gif)
